baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/OptionalOrElseGetConstant.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.errorprone;
+
+import com.google.auto.service.AutoService;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.BugPattern.SeverityLevel;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.bugpatterns.BugChecker.MethodInvocationTreeMatcher;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.fixes.SuggestedFixes;
+import com.google.errorprone.matchers.CompileTimeConstantExpressionMatcher;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.method.MethodMatchers;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.LambdaExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.Tree;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        name = "OptionalOrElseGetConstant",
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        providesFix = BugPattern.ProvidesFix.REQUIRES_HUMAN_ATTENTION,
+        severity = SeverityLevel.ERROR,
+        summary = "If lambda passed to Optional#orElseGet returns a compile time constant, use Optional#orElse instead")
```

Comment:
I think we should expand this check to include any variable

Reply:
+1 no need for it to be constant, field loads are less expensive than lambda allocations.

We would want to accept `foo`, `this.foo`, `foo.bar` but not `foo().bar`, `foo[index]`, `foo[index].bar`.

Reply:
Was a bit worried about race conditions in case it's a member variable. Thinking of it now again though, I'm realizing that it's probably not an issue. I'll look into it.

Reply:
Out of curiosity, why not `foo[index]` and `foo[index].bar`?

Reply:
I'm not entirely sure they'll always be better, but they probably are. Even if there's llc miss it's probably better than the lambda.
I tend to prefer flagging and fixing the most unambiguous cases, but I don't have a strong opinion here.

Reply:
Check now covers `foo`, `this.foo`, `foo.bar` but not array accesses or anything that could have side-effects.